src/java.base/share/classes/java/lang/foreign/Linker.java

```diff
@@ -241,30 +241,36 @@
  * </tbody>
  * </table></blockquote>
  * <p>
- * All native linker implementations support a well-defined subset of layouts. More formally,
- * a layout {@code L} is supported by a native linker {@code NL} if:
+ * A native linker only supports function descriptors whose argument/return layouts are
+ * <em>well-formed</em> layouts. More formally, a layout `L`is well-formed if:
  * <ul>
- * <li>{@code L} is a value layout {@code V} and {@code V.withoutName()} is a canonical layout</li>
+ * <li>{@code L} is a value layout and {@code L} is derived from a canonical layout
+ *     {@code C} such that {@code L.byteAlignment() <= C.byteAlignment()}</li>
  * <li>{@code L} is a sequence layout {@code S} and all the following conditions hold:
  * <ol>
- * <li>the alignment constraint of {@code S} is set to its
- *     <a href="MemoryLayout.html#layout-align">natural alignment</a>, and</li>
- * <li>{@code S.elementLayout()} is a layout supported by {@code NL}.</li>
+ * <li>{@code L.byteAlignment()} is equal to the sequence layout's <em>natural alignment</em>
+ *     , and</li>
+ * <li>{@code S.elementLayout()} is a well-defined layout.</li>
  * </ol>
  * </li>
  * <li>{@code L} is a group layout {@code G} and all the following conditions hold:
  * <ol>
- * <li>the alignment constraint of {@code G} is set to its
- *     <a href="MemoryLayout.html#layout-align">natural alignment</a>;</li>
- * <li>the size of {@code G} is a multiple of its alignment constraint;</li>
- * <li>each member layout in {@code G.memberLayouts()} is either a padding layout or
- *     a layout supported by {@code NL}, and</li>
- * <li>{@code G} does not contain padding other than what is strictly required to align
- *      its non-padding layout elements, or to satisfy (2).</li>
+ * <li>{@code G.byteAlignment()} is equal to the group layout's <em>natural alignment</em></li>
+ * <li>{@code G.byteSize()} is a multiple of {@code G.byteAlignment()}</li>
+ * <li>Each member layout in {@code G.memberLayouts()} is either a padding layout or a
+ *     well-defined layout</li>
+ * <li>Each non-padding member layout {@code E} in {@code G.memberLayouts()} follows an
+ *     optional padding member layout, whose size is the minimum size required to
+ *     align {@code E}</li>
+ * <li>{@code G} contains an optional trailing padding member layout, whose size is the
+ *     minimum size that satisfies (2)</li>
  * </ol>
  * </li>
  * </ul>
  *
+ * Linker implementations may optionally impose additional platform-specific constraints
```

Comment:
I think this sentence needs to be better joined with the next. Because here we now say that we can reject more stuff -- but the para after still says that a Linker can accept more stuff. We should basically say that some layouts, such as packed struct layouts, even if well-formed, might be rejected by the Linker.

In other words we should make it clear that being well-formed is a necessary condition, not a sufficient one. Unfortunately attempting to capture a sufficient condition is problematic (what we did before), because that has too much platform-dependency variations. But this should be spelled out somehow, because it's easy, when reading the doc, to think that the set of rules above are both necessary and sufficient (e.g. if well-formed then is accepted by Linker and vice-versa) which is a big part of the discussion that happened in this PR. The main point to drive home is that some rules cannot be fully specified as they vary too much.